### PR TITLE
[Enhancement]optimize subfield expr evaluate when copy is no need

### PR DESCRIPTION
**Pull request by zombee0**
Why I'm doing:
subfield expr of struct evaluate cost a lot because of memory copy.

What I'm doing:
Don't copy when copy is no need for subfiled expr.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExprNoCopyRule.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        // project expression
+        LogicalProjectOperator project = input.getOp().cast();
+        List<ScalarOperator> projectMapValues = new ArrayList<>(project.getColumnRefMap().values());
+        for (int i = 0; i < projectMapValues.size(); i++) {
+            ScalarOperator value = projectMapValues.get(i);
+            // only deal with subfield expr of slotRef
+            if (value instanceof SubfieldOperator && value.getChild(0) instanceof ColumnRefOperator) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExprNoCopyRule.java:50]: I think don't need the required?

File: gensrc/thrift/Exprs.thrift [hunk 2]
```diff
@@ -222,6 +222,7 @@ struct TExprNode {
   // Used for SubfieldExpr
   30: optional list<string> used_subfield_names;
   31: optional TBinaryLiteral binary_literal;
+  32: optional bool copy_flag;
```
Comment (Smith-Cruise) [inline gensrc/thrift/Exprs.thrift:225]: add comment this flag only used for subfield expr

**Review by Smith-Cruise — APPROVED** [review]

File: gensrc/thrift/Exprs.thrift (hunk 2 above)
Comment (zombee0) [reply gensrc/thrift/Exprs.thrift:225]: will do it

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExprNoCopyRule.java (hunk 1 above)
Comment (zombee0) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldExprNoCopyRule.java:50]: this is used to simplify the case

**Review by Seaven — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**